aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -481,6 +626,118 @@ module aptos_framework::delegation_pool {
+            delegate_voting_power_events: account::new_event_handle<DelegateVotingPowerEvent>(&stake_pool_signer),
+        });
+    }
+
+    public entry fun vote(voter: &signer, pool_address: address, proposal_id: u64, voting_power: u64, should_pass: bool) acquires DelegationPool, GovernanceRecords {
+        assert_partial_governance_voting_enabled(pool_address);
+
+        // synchronize delegation and stake pools before any user operation
+        synchronize_delegation_pool(pool_address);
+
+        let voter_address = signer::address_of(voter);
+        let remaining_voting_power = calculate_and_update_remaining_voting_power(pool_address, voter_address, proposal_id);
+        if (voting_power > remaining_voting_power) {
+            voting_power = remaining_voting_power;
+        };
+        assert!(voting_power > 0, error::invalid_argument(ENO_VOTING_POWER));
+
+        let governance_records = borrow_global_mut<GovernanceRecords>(pool_address);
+        let stake_pool_remaining_voting_power = aptos_governance::get_remaining_voting_power(pool_address, proposal_id);
+        let stake_pool_used_voting_power = aptos_governance::get_voting_power(pool_address) - stake_pool_remaining_voting_power;
+        let proposal_used_voting_power = smart_table::borrow_mut_with_default(&mut governance_records.votes_per_proposal, proposal_id, 0);
+        // A edge case: Before enabling partial governance voting on a delegation pool, the delegation pool has
+        // a voter which can vote with all voting power of this delegation pool. If the voter votes on a proposal after
+        // partial governance voting flag is enabled, the delegation pool doesn't have enough voting power on this
+        // proposal for all the delegators. To be fair, no one can vote on this proposal through this delegation pool.
+        // To detect this case, check if the stake pool had used voting power not through delegation_pool module.
+        assert!(stake_pool_used_voting_power == *proposal_used_voting_power, error::invalid_argument(EALREADY_VOTED_BEFORE_ENABLE_PARTIAL_VOTING));
+        *proposal_used_voting_power = *proposal_used_voting_power + voting_power;
+        let used_voting_power = borrow_mut_used_voting_power(governance_records, voter_address, proposal_id);
+        *used_voting_power = *used_voting_power + voting_power;
+
+        let pool_signer = retrieve_stake_pool_owner(borrow_global<DelegationPool>(pool_address));
+        aptos_governance::partial_vote(&pool_signer, pool_address, proposal_id, voting_power, should_pass);
+
+        event::emit_event(
+            &mut governance_records.vote_events,
+            VoteEvent {
+                voter: voter_address,
+                proposal_id,
+                delegation_pool: pool_address,
+                num_votes: voting_power,
+                should_pass,
+            }
+        );
+    }
+
+    public entry fun create_proposal(
```

Comment:
Can you add docs for all public functions? May be something concise if there's not a lot to explain (i.e. the function is self-explaining)